### PR TITLE
add "live tests"

### PR DESCRIPTION
**Pull request by davepacheco**
(some of this text is copied from the new README)

This PR adds a new Omicron package called `omicron-live-tests` to contain automated tests that operate in the context of an already-deployed "real" Oxide system (e.g., `a4x2` or our `london` or `madrid` test environments).  The motivation is to provide a home for automated tests for all kinds of Reconfigurator behavior (e.g., add/expunge of all zones, add/expunge sled, upgrades, etc.), though it could be used for non-Reconfigurator behavior, too.

What makes these tests different from the rest of the test suite is that they require connectivity to the underlay network of the deployed system and they make API calls to various components in that system and they assume that this will behave like a real production system.  By contrast, the normal tests instead _set up_ a bunch of components using simulated sled agents and localhost networking, which is great for starting from a predictable state and running tests in parallel, but the simulated sled agents and networking make it impossible to exercise quite a lot of Reconfigurator's functionality.  

There are some safeguards so that these tests won't run on production systems: they refuse to run if they find any Oxide-hardware sleds in the system whose serial numbers don't correspond to known test environments.

This is similar to `end-to-end-tests` in a lot of ways but I didn't think it made sense to use the same package because the test environment is pretty different.  The end-to-end-tests run on a Helios system on which we have deployed Sled Agent and a bunch of components in zones.  But there aren't multiple sleds and there isn't a real underlay network.  I don't think it's faithful enough to carry out a lot of the Reconfigurator tests that we'd like to do.

Like the end-to-end-tests, this package is _not_ built or tested by default because the tests generally can't work in a dev environment and there's no way to have `cargo` build and check them but not run the tests by default.

Eventually I hope we can find a way to run these tests automatically in CI, but that's future work.  For now, there are instructions for running these by hand on an Omicron system.  Start by running `cargo xtask live-tests` to build an archive and then follow the instructions:

```
$ cargo xtask live-tests
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.96s
     Running `target/debug/xtask live-tests`
using temporary directory: /dangerzone/omicron_tmp/.tmp0ItZUD
will create archive file:  /dangerzone/omicron_tmp/.tmp0ItZUD/live-tests-archive/omicron-live-tests.tar.zst
output tarball:            /home/dap/omicron-work/target/live-tests-archive.tgz

running: /home/dap/.rustup/toolchains/1.80.1-x86_64-unknown-illumos/bin/cargo "nextest" "archive" "--package" "omicron-live-tests" "--archive-file" "/dangerzone/omicron_tmp/.tmp0ItZUD/live-tests-archive/omicron-live-tests.tar.zst"
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.89s
info: experimental features enabled: setup-scripts
   Archiving 1 binary, 1 build script output directory, and 1 linked path to /dangerzone/omicron_tmp/.tmp0ItZUD/live-tests-archive/omicron-live-tests.tar.zst
    Archived 35 files to /dangerzone/omicron_tmp/.tmp0ItZUD/live-tests-archive/omicron-live-tests.tar.zst in 0.31s
running: bash "-c" "tar cf - Cargo.toml .config/nextest.toml live-tests | tar xf - -C \"/dangerzone/omicron_tmp/.tmp0ItZUD/live-tests-archive\""
running: tar "cf" "/home/dap/omicron-work/target/live-tests-archive.tgz" "-C" "/dangerzone/omicron_tmp/.tmp0ItZUD" "live-tests-archive"
created: /home/dap/omicron-work/target/live-tests-archive.tgz

To use this:

1. Copy the tarball to the switch zone in a deployed Omicron system.

     e.g., scp \
              /home/dap/omicron-work/target/live-tests-archive.tgz \
              root@YOUR_SCRIMLET_GZ_IP:/zone/oxz_switch/root/root

2. Copy the `cargo-nextest` binary to the same place.

     e.g., scp \
              $(which cargo-nextest) \
              root@YOUR_SCRIMLET_GZ_IP:/zone/oxz_switch/root/root

3. On that system, unpack the tarball with:

     tar xzf live-tests-archive.tgz

4. On that system, run tests with:

     TMPDIR=/var/tmp ./cargo-nextest nextest run \
         --archive-file live-tests-archive/omicron-live-tests.tar.zst \
         --workspace-remap live-tests-archive
```

Follow the instructions, run the tests, and you'll see the usual `nextest`-style output:

```
root@oxz_switch:~# TMPDIR=/var/tmp ./cargo-nextest nextest run          --archive-file live-tests-archive/omicron-live-tests.tar.zst          --workspace-remap live-tests-archive
  Extracting 1 binary, 1 build script output directory, and 1 linked path to /var/tmp/nextest-archive-Lqx9VZ
   Extracted 35 files to /var/tmp/nextest-archive-Lqx9VZ in 1.01s
info: experimental features enabled: setup-scripts
    Starting 1 test across 1 binary (run ID: a5fc9163-9dd5-4b23-b89f-55f8f39ebbbc, nextest profile: default)
        SLOW [> 60.000s] omicron-live-tests::test_nexus_add_remove test_nexus_add_remove
        PASS [  61.975s] omicron-live-tests::test_nexus_add_remove test_nexus_add_remove
------------
     Summary [  61.983s] 1 test run: 1 passed (1 slow), 0 skipped
root@oxz_switch:~# 
```

### CODE REVIEW

**Review by jgallagher — APPROVED** [review]
Looks great! Just a few questions

File: .config/nextest.toml [hunk 1]
```diff
@@ -10,7 +10,7 @@ experimental = ["setup-scripts"]
 [[profile.default.scripts]]
 # Exclude omicron-dev tests from crdb-seed as we explicitly want to simulate an
 # environment where the seed file doesn't exist.
```
Comment (sunshowers) [inline .config/nextest.toml:12]: Could you also expand this comment to cover `omicron-live-tests`?

File: .config/nextest.toml [hunk 2]
```diff
@@ -21,18 +21,26 @@ fail-fast = false
...
 # the nodes in the cluster. We would like to relax this in the future, at which
 # point this test-group configuration can be removed or at least loosened to
 # support testing in parallel. For now, enforce strict serialization for all
 # tests with `replicated` in the name.
-[test-groups]
 clickhouse-cluster = { max-threads = 1 }
+# While most Omicron tests operate with their own simulated control plane, the
+# live-tests operate on a more realistic, shared control plane and test
+# behaviors that conflict with each other.  They need to be run serially.
+live-tests = { max-threads = 1 }
```
Comment (sunshowers) [inline .config/nextest.toml:34]: 1. Makes sense!
2. As mentioned in the DM, you can now filter out `omicron-live-tests` from [the default set](https://nexte.st/docs/running/#running-a-subset-of-tests-by-default). (You'll want to bump the required and recommended versions in this file, as well as the version used in CI, to 0.9.76)

File: dev-tools/xtask/src/live_tests.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+pub fn run_cmd(_args: Args) -> Result<()> {
+    const NAME: &str = "live-tests-archive";
+
+    // The live tests operate in deployed environments, which always run
+    // illumos.  Bail out quickly if someone tries to run this on a system whose
+    // binaries won't be usable.  (We could compile this subcommand out
+    // altogether on non-illumos systems, but it seems more confusing to be
+    // silently missing something you might expect to be there.  Plus, you can
+    // still check and even build *this* code on non-illumos systems.)
+    if cfg!(not(target_os = "illumos")) {
```
Comment (sunshowers) [inline dev-tools/xtask/src/live_tests.rs:24]: Interesting, this makes sense.

File: dev-tools/xtask/src/live_tests.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        bail!("live-tests archive can only be built on illumos systems");
+    }
+
+    let tmpdir_root =
+        camino_tempfile::tempdir().context("creating temporary directory")?;
+    let final_tarball = camino::Utf8PathBuf::try_from(
+        std::env::current_dir()
```
Comment (sunshowers) [inline dev-tools/xtask/src/live_tests.rs:31]: Hmm I think this is common enough that I should add this to camino...

File: live-tests/tests/common/mod.rs [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    if let Some(first) = tmpdir_components.next() {
+        if let Some(next) = tmpdir_components.next() {
+            if first == Component::RootDir
+                && next == Component::Normal(OsStr::new("tmp"))
+            {
+                eprintln!(
+                    "WARNING: temporary directory appears to be under /tmp, \
+                     which is generally tmpfs.  Consider setting \
+                     TMPDIR=/var/tmp to avoid runaway tests using too much\
+                     memory and swap."
```
Comment (sunshowers) [inline live-tests/tests/common/mod.rs:190]: Should we check for a minimum amount of free space required for `TMPDIR`?

File: live-tests/tests/common/mod.rs [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        anyhow!("{}", textwrap::wrap(&text, 80).join("\n"))
+    })?;
+
+    // Warn the user if the temporary directory is /tmp.  This check is
+    // heuristic.  There are other ways they may have specified a tmpfs
+    // temporary directory and we don't claim to catch all of them.
+    //
+    // We could also just go ahead and use /var/tmp, but it's not clear we can
+    // reliably do that at this point (if Rust or other components have cached
+    // TMPDIR) and it would be hard to override.
```
Comment (sunshowers) [inline live-tests/tests/common/mod.rs:178]: Interesting, you could use a setup script to set `TMPDIR` outside of the test process.

**Review by sunshowers — COMMENTED** [review]
This is so cool, thanks for doing it!

File: live-tests/tests/test_nexus_add_remove.rs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+use omicron_test_utils::dev::poll::CondCheckError;
+use slog::{debug, info};
+use std::net::SocketAddrV6;
+use std::time::Duration;
+
+// TODO-coverage This test could check other stuff:
+//
+// - that after adding:
+//   - the new Nexus appears in external DNS
+//   - we can _use_ the new Nexus from the outside
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply live-tests/tests/test_nexus_add_remove.rs:31]: Possibly not today.  I suspect figuring this out will be worth the investment.  Maybe we can use the techport interface for the external API?

File: live-tests/tests/common/mod.rs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+/// to run this in dogfood or other pre-production or production environments.
+/// This function uses an allowlist of Oxide serials corresponding to test
+/// environments so that it never accidentally runs on a production system.
+///
+/// Non-Oxide hardware (e.g., PCs, a4x2, etc.) are always allowed.
+async fn check_hardware_environment(
+    opctx: &OpContext,
+    datastore: &DataStore,
+) -> Result<(), anyhow::Error> {
+    const ALLOWED_GIMLET_SERIALS: &[&str] = &[
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply live-tests/tests/common/mod.rs:212]: It could.  For what it's worth, when I touch this file and rebuild, it takes 13.5s.  You do have the extra copying step(s) that are also annoying.

File: live-tests/tests/common/mod.rs (hunk 5 above)
Comment (davepacheco) [reply live-tests/tests/common/mod.rs:190]: I don't think so.  Management of tmp space seems outside the scope of all this.  I almost didn't bother with this warning at all except that the failure mode of running out of space in /tmp (i.e., swap) is usually pretty bad for the whole zone, and that could be pretty annoying to clean up in the switch zone.  That said, I'm barely pro for even this warning -- I could be convinced to just remove this altogether.

Also: there are so many ways I've seen that kind of check be wrong, in part because it's _really_ hard to know how much physical disk space something is going to need before it writes it.  Example: false positive errors when using zfs with compression=on because a thing thinks it needs a GiB even though after compression it'll only use 150 MiB of space.  Plus other things can use space in the meantime or space can free up in the meantime.

File: live-tests/tests/common/mod.rs (hunk 6 above)
Comment (davepacheco) [reply live-tests/tests/common/mod.rs:178]: Does the test process inherit the environment of the setup script?

File: .config/nextest.toml (hunk 1 above)
Comment (davepacheco) [reply .config/nextest.toml:12]: Done in c7cd0ed63.

File: .config/nextest.toml (hunk 2 above)
Comment (davepacheco) [reply .config/nextest.toml:36]: This is cool.  I'd prefer we do that in a separate PR.

File: live-tests/tests/common/mod.rs (hunk 6 above)
Comment (sunshowers) [reply live-tests/tests/common/mod.rs:178]: Yes, if you write it out to `$NEXTEST_ENV`: https://nexte.st/docs/configuration/setup-scripts/#environment-variables

(This is how the crdb-seed setup script works)

File: live-tests/tests/test_nexus_add_remove.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jgallagher) [reply live-tests/tests/test_nexus_add_remove.rs:31]: We could, but I'm not sure it gains us much. If we're in the switch zone we can either connect to the TCP proxy wicketd runs or directly to the "Nexus external API exposed on the underlay" that wicketd forwards to. But in either case, it wouldn't really confirm external connectivity, because it's not actually using the external IP. It does confirm the new Nexus would be reachable via the techport, which is something though.

**Comment by davepacheco** [comment]
I was hoping to land this at a point where the live tests built with this branch passed against a deployment built from the tip of this branch.  But I can't keep up with "main" today -- it's moving faster than I can iterate.  I have run the live tests built with ab5448b57d9266ac247f46c69bbcdc29ca80ae1c (current tip of this PR) against e85d012e80d4c83078c71cd7aee14d6b051c1bb6 from this branch, which is approximately main at 758818aea.  Although that's only about 6 hours old, it's 8 commits behind.

**Comment by davepacheco** [comment]
Okay, from the tip of this branch, locally, I merged with main @ 3dcf1e766c4d44c56df5ce80f3180fbbea00de4a, deployed a system with that, and ran the live-tests with that and that worked, too.

File: live-tests/macros/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+/// ```ignore
+/// #[live_test]
+/// async fn test_my_test_case(lc: &mut LiveTestContext) {
+///   assert!(true);
+/// }
+/// ```
+///
+/// We use this instead of implementing Drop on LiveTestContext because we want
+/// the teardown to only happen when the test doesn't fail (which causes a panic
+/// and unwind).
```
[the remark this answers is not in the document]
Comment (davepacheco) [reply live-tests/macros/src/lib.rs:33]: I went down this path in a follow-on branch:
https://github.com/oxidecomputer/omicron/pull/new/dap/experiments/live-tests-no-macro

A few notes:

- I wanted to avoid type parameters in `LiveTestContext::run_test` because I didn't want it monomorphized for every test that we add.  So I had `run_test` accept a `&dyn Fn` instead of `F: Fn` and that closure returns a `BoxFuture`.
- rustfmt seems to have lost its mind on the code and it looks awful.  I have spent no time trying to figure out why.
- It doesn't quite compile due to a lifetime issue.  I gave up (not at all insurmountable, I was just spending more time on it than I had hoped to for just an experiment).

There are a couple of problems:

- the rustfmt problem, but I'll assume that is surmountable
- it's much easier to copy/paste the code and get the wrong `test_name` in the argument
- `[tokio::test]` and so `[live_test]` allow your test function to return nothing at all or a Result type.  I had this version return a `Result`.  It would be tricky to support both here.  (But most of our tests probably just return `()` and we could probably just have it support that.)

In summary, I don't like it better but I'm still open to it if people really do!

[the remark this answers is not in the document]
Comment (sunshowers) [reply live-tests/macros/src/lib.rs:33]: Oh thanks! Looking at https://github.com/oxidecomputer/omicron/commit/3f43f781e154d8c888b146947a83b91370adff33:

* The rustfmt issue is because of the additional nesting involved afaict.
* I think monomorphizing that is okay? The function looks pretty small to me, and the dependent functions aren't generic.
3. Passing in borrowed data to closures is difficult. [async closures](https://github.com/rust-lang/rust/issues/62290) solve this but aren't stable yet. I ran into this with the update engine, and that has to pass an owned context in for that reason.
